src/librustc_resolve/late.rs

```diff
@@ -1790,6 +1815,20 @@ impl<'a, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
         if let Some(LexicalScopeBinding::Res(res)) = binding { res != Res::Err } else { false }
     }
 
+    /// A wrapper around [`Resolver::report_error`].
+    ///
+    /// This doesn't emit errors for function bodies if `ignore_bodies` is set.
```

Comment:
This comment seems slightly misleading; its not as if the implementation of this helper *knows* that its being called from within the context of a function body or not, right? That is instead a property of the call sites. So I'd just write "This doesn't emit errors if` ignore_bodies` is set".

That then leads me to wonder; are all of the current call sites always within function bodies, or are there some that are reporting errors from name resolutions from e.g. type signatures? (Its an honest question; I don't know the answer one way or another.) If there's a simple way to re-enforce that expectation, such as naming this `report_error_within_fn_body`, that might help...

Reply:
> its not as if the implementation of this helper knows that its being called from within the context of a function body or not, right?

It does know actually - if you look at `should_report_errs()` that checks whether `current_function.is_none()` and if so says the error should be reported.

I'm not sure what the behavior is if there is an impl nested inside a function - it should give an error if not resolved but that might not be the current behavior.

Reply:
Ok yes this does not have the proper behavior for nested impls.
```rust
$ cat impl-fn-nesting.rs 
struct MyType;

fn f() {
    impl MyTrait for MyType {}
}
$ rustdoc +stable impl-fn-nesting.rs 
error[E0405]: cannot find trait `MyTrait` in this scope
 --> impl-fn-nesting.rs:4:10
  |
4 |     impl MyTrait for MyType {}
  |          ^^^^^^^ not found in this scope

error: Compilation failed, aborting rustdoc

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0405`.
$ rustdoc +stage1 impl-fn-nesting.rs 
$
```
Maybe errors about trait implementations and macros should be special cased so that they still give an error? The criteria here is whether the item is publicly visible.

Reply:
I added some special casing and this now correctly gives an error but I'm not sure how well the approach generalizes ... it certainly doesn't handle `fn -> impl -> fn` (test case in latest commit) and there may be more things I didn't think of.

Reply:
I think I need to look at `self.ribs` instead of blanket 'in trait' and 'in function' lookups, that might get me the inner-most scope.

Reply:
I added a commit which fixes this case. Not sure I thought of all the error cases, but getting closer.

Reply:
Ok, got it for real now. Instead of trying to reconstruct the state from whatever misc info was lying around at the time, I just added an explicit variable on `self` to keep track of it. This is ready for re-review (let me know if I should squash the commits, most are just me debugging).

> This comment seems slightly misleading; its not as if the implementation of this helper knows that its being called from within the context of a function body or not, right? That is instead a property of the call sites.

Exactly, so I added information to `self.ignore_bodies_state` to keep track of where this was being called from. It works as a stack but only needs to updated in a few select places.